contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,315 @@
+
+Currently, cAdvisor collects a large number of metrics related to system and container performance. However, only some of these metrics are consumed by the kubelet summary API, and many are not used.  The kubelet summary API is published to the kubelet summary API endpoint.  Some of the metrics provided by the summary API are consumed internally, but most are not used internally.
+
+### Motivations
+
+Giving the kubelet the role of both providing metrics for its own use, and providing metrics for users has a couple problems
+ - First, it is clearly inefficent to collect metrics and not use them.  The kubelet uses only a small portion of the metrics it collects, and thus presents considerable extra overhead to any users who do not use them, or prefer a third party monitoring solution.
+ - Second, as the number of metrics collected grows over time, the kubelet will gain more and more overhead for collecting, processing, and publishing these metrics.  Since the metrics users may want is unbounded, the kubelet's resource overhead could easily grow to unreasonable levels.
+
+It is very cumbersome to make changes or bugfixes in cAdvisor, because that then needs to be vendored back into kubernetes.
+
+CAdvisor is structured to collect metrics on an interval, which is appropriate for a stand-alone metrics collector.  However, many functions in the kubelet are latency-sensitive (eviction, for example), and would benifit from a more "On-Demand" metrics collection design.
+
+### Proposal
+
+I propose to create a set of core metrics, collected by the kubelet, and used solely by internal kubernetes components.
+
+### Non Goals
+Everything covered in the [Monitoring Architecture](https://github.com/kubernetes/kubernetes/blob/master/docs/design/monitoring_architecture.md) design doc.  This includes the third party metrics pipeline, and the methods by which the metrics found in this proposal are provided to other kubernetes components.
+
+Integration with CRI will not be covered in this proposal.  In future proposals, integrating with CRI may provide a better abstraction of information required by the core metrics pipeline to collect metrics.
+
+## Design
+
+This design covers only the internal Core Metrics Pipeline.
+
+High level requirements for the design are as follows:
+ - Do not break existing users.  We should continue to provide the full summary API by default.
```

Comment:
I think @vishh is saying that you should qualify this, e.g.:

> We should continue to provide the fully summary API as an optional addon.  Once the monitoring pipeline is converted to use the new API, the summary API will be moved out into a component which can optionally be served using a standalone copy of cAdvisor

or something to that effect.